rcl_yaml_param_parser/src/parser.c

```diff
@@ -187,7 +187,14 @@ static rcl_ret_t add_name_to_ns(
       ns_len = strlen(cur_ns);
       name_len = strlen(name);
       sep_len = strlen(sep_str);
-      tot_len = ns_len + name_len + sep_len + 1U;
+      // Check the last sep_len characters of the current NS against the separator string.
+      if (strcmp(cur_ns + ns_len - sep_len, sep_str) == 0) {
+        // Current NS already ends with the separator: don't put another separator in.
```

Comment:
Looking at the tests I think the intent was that each token in the namespace would be it's own level in the config. Namespace `/foo/bar/baz` and node name `smith` would be described by:

```
foo:
  bar:
    baz:
      smith:
        ros__parameters:
```

I'm guessing it was not expected that the names could contain `/` and `.`. Allowing them does look more convenient though. To support that there is at least one more bug ros2/rcl#301.

Reply:
by the way, if we're not sure if we want to support the case where there are trailing slashes _in general_ we can modify the change in this commit d311e0c55294bc4a7b1e2e8e621b2d0e886a5a78 and go back to just making the exception for the root namespace (maybe error in other cases).